curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -1331,6 +1586,54 @@ void DataCache::Release() {
     return;
 }
 
+void DataCache::CopyDataCacheToData(uint64_t offset, uint64_t len, char* data) {
+    assert(offset + len <= len_);
+    uint64_t blockSize = s3ClientAdaptor_->GetBlockSize();
+    uint32_t pageSize = s3ClientAdaptor_->GetPageSize();
+    uint64_t newChunkPos = chunkPos_ + offset;
+    uint64_t blockIndex = newChunkPos / blockSize;
+    uint64_t blockPos = newChunkPos % blockSize;
+    uint64_t pagePos, pageIndex;
+    uint64_t n, m, blockLen;
+    uint64_t dataOffset = 0;
+
+    VLOG(9) << "CopyDataCacheToData start dataOffset:" << dataOffset
+            << ",len:" << len;
+    while (len > 0) {
+        if (blockPos + len > blockSize) {
+            n = blockSize - blockPos;
+        } else {
+            n = len;
+        }
+        blockLen = n;
+        PageDataMap& pdMap = dataMap_[blockIndex];
+        PageData* pageData = NULL;
+        pageIndex = blockPos / pageSize;
+        pagePos = blockPos % pageSize;
+        while (blockLen > 0) {
+            if (pagePos + blockLen > pageSize) {
+                m = pageSize - pagePos;
+            } else {
+                m = blockLen;
+            }
+
+            assert(pdMap.count(pageIndex));
+            pageData = pdMap[pageIndex];
+            memcpy(data + dataOffset, pageData->data + pagePos, m);
+            pageIndex++;
+            blockLen -= m;
+            dataOffset += m;
+            pagePos = (pagePos + m) % pageSize;
+        }
+
+        blockIndex++;
+        len -= n;
+        blockPos = (blockPos + n) % blockSize;
+    }
+    VLOG(9) << "CopyDataCacheToData end.";
+    return;
+}
+
```

Comment:
for `Flush` function below, I think you should add another policy to avoid uploading small objects to s3.
e.g., if `Flush` is not triggered by user's write, you should skip uploading the current cache if its length is too small.